### ISSUE TITLE
Quote passwords when passing them to subscription manager

### ISSUE THREAD

**Issue by dustymabe** [minishift/minishift#570]
Passwords that have special characters in them need to be quoted on the command line (if you run it through a shell which will interpret such characters). We either need to quote the password with `''` or we need to not run the command through a shell, which may be preferable.

Either way, once a fix is proposed, please try to run through the process with some special characters defined (`%$^`, etc) and make sure you can log in.

**Comment by dustymabe** [comment]
FYI - if this is not already fixed (maybe I missed it somewhere), then it should probably be fixed before next CDK release.

**Comment by hferentschik** [comment]
@dustymabe, just to better understand the problem here. You are using the latest CDK 3 beta and try to register it by passing your subscription credentials during the start command. You are having a password with special characters which you escape accordingly on the command line when executing the start command. However, this will still fail to start the CDK, since we are internally not passing the password quoted to the subscription manager? Is this correct? 

Maybe you could provide an example command and the error you are seeing?

BTW, since this seems to be CDK issue, it should actually be tracked in the CDK JIRA issue tracker - https://issues.jboss.org/projects/CDK. Any change you can open an issue there?

**Comment by dustymabe** [comment]
> @dustymabe, just to better understand the problem here. You are using the latest CDK 3 beta and try to register it by passing your subscription credentials during the start command. You are having a password with special characters which you escape accordingly on the command line when executing the start command. However, this will still fail to start the CDK, since we are internally not passing the password quoted to the subscription manager? Is this correct?

that is correct

> Maybe you could provide an example command and the error you are seeing?

i run minishift start with my user/pass in env vars and it fails

```
[cloud-user@cloudhost ~]$ minishift start
Starting local OpenShift cluster using 'kvm' hypervisor...
Registering machine using subscription-manager
E0319 20:21:38.506799   18203 start.go:150] Error starting the VM: Error registering the VM: ssh command error:
command : sudo subscription-manager register --auto-attach --username yourusername --password pa$$word 
err     : exit status 70
output  : Invalid username or password. To create a login, please visit https://www.redhat.com/wapps/ugc/register.html
Registering to: subscription.rhsm.redhat.com:443/subscription
. Retrying.
```

> BTW, since this seems to be CDK issue, it should actually be tracked in the CDK JIRA issue tracker - https://issues.jboss.org/projects/CDK. Any change you can open an issue there?

I logged in with my red hat customer portal account and it asks me to create a new JBOSS developer account. This is a barrier to entry I don't think many people are going to jump through.

**Comment by dustymabe** [comment]
note, I have verified that if I run the subscription-manager command manually with single quotes like `'pa$$word'`, it works. I also will note that I changed my password to not have special characters in it and I do not have this problem with minishift.

**Comment by hferentschik** [comment]
I created https://issues.jboss.org/browse/CDK-100 to track this in the CDK issue tracker.

**Comment by hferentschik** [comment]
> note, I have verified that if I run the subscription-manager command manually with single quotes like 'pa$$word', it works. I also will note that I changed my password to not have special characters in it and I do not have this problem with minishift.

Cool. Quoting will work, but only for the cases where the password does not a single quote either. Not running via a shell might not be an option, since we need to source the environment which potentially provides proxy settings.

**Comment by LalatenduMohanty** [comment]
@dustymabe @hferentschik I have `!!` in my RHN password and worked for me for CDK 3 Beta. Not sure if my Linux shell have something which makes it work.

**Comment by hferentschik** [comment]
@LalatenduMohanty You literally mean '!'? I don't think this is an issue. Biggest issue are most likely quotes. The problem is here https://github.com/minishift/minishift/blob/e2ba93544af908b8f889593cad4d1e300e23f7f9/pkg/minishift/registration/redhat.go#L58-L58. See how we pass the password unquoted to subscription manager?

**Comment by dustymabe** [comment]
I would say that the biggest issue is probably characters that the shell interprets as special characters. Shell uses `$` to expand variables, so that is the prime candidate. @LalatenduMohanty - try putting a `$` in your password and see if you observe the issue.

**Comment by hferentschik** [comment]
Merged via pull request #672
